crypto/stark-verifier-ethereum/contracts/fri.sol

```diff
@@ -29,20 +83,203 @@ contract Fri {
+        uint8 log_eval_domain_size,
+        uint64[] memory queries,
+        uint256[] memory polynomial_at_queries,
+        uint256 oods_point,
+        uint256 evaluated_oods_point
+    ) internal {
+        fold_and_check_fri_layers(
+            FriContext(
+                proof.fri_values,
+                proof.fri_commitments,
+                proof.fri_decommitments,
+                fri_layout,
+                eval_points,
+                log_eval_domain_size,
+                queries,
+                polynomial_at_queries,
+                proof.last_layer_coeffiencts
+            )
+        );
+
+        // The final check is that the constraints evaluated at the out of domain sample are
+        // equal to the values commited constraint values
+        uint256 result = 0;
+        uint256 power  = 0x0000000000000000000000000000000000000000000000000000000000000001.to_montgomery();
+        for (uint256 i = 0; i < proof.constraint_oods_values.length; i++) {
+            result = result.fadd(proof.constraint_oods_values[i].fmul_mont(power));
+            power = power.fmul_mont(oods_point);
+        }
+        require(result == evaluated_oods_point, "Oods mismatch");
+    }
+
+    // This function takes in fri values, decommitments, and layout and checks the folding and merkle proofs
+    // Note the final layer folded values will live in the
+    function fold_and_check_fri_layers(FriContext memory fri_data) internal {
+        Eval_X memory eval = init_eval(fri_data.log_eval_domain_size);
+        LayerContext memory layer_context = LayerContext({
+            len: uint64(2)**(fri_data.log_eval_domain_size),
+            step: 1,
+            coset_size: 0
+        });
+        uint256[] memory merkle_ind = new uint256[](fri_data.queries.length);
+        bytes32[] memory merkle_val = new bytes32[](fri_data.queries.length);
+
+        for (uint256 i = 0; i < fri_data.fri_layout.length; i++) {
+            layer_context.coset_size = uint64(2)**(fri_data.fri_layout[i]);
+            // Overwrites and resizes the data array and the querry index array
+            // They will contain the folded points and indexes
+            fold_layer(
+                fri_data.polynomial_at_queries,
+                fri_data.queries,
+                Iterators.init_iterator(fri_data.fri_values[i]),
+                eval,
+                fri_data.eval_points[i],
+                layer_context,
+                merkle_val
+            );
+            // Merkle verification is in place but we need unchanged data in the next loop.
+            fri_data.queries.deep_copy_and_convert(merkle_ind);
+            // Since these two arrays only shrink we can safely resize them
+            if (fri_data.queries.length != merkle_ind.length) {
+                uint256 num_queries = fri_data.queries.length;
+                assembly {
+                    mstore(merkle_ind, num_queries)
+                    mstore(merkle_val, num_queries)
+                }
```

Comment:
Can we avoid assembly here? I vaguely remember an assignment working: `merkle_ind.length = num_queries` working. If not, we can use `num_queries` instead of length.

Reply:
They removed that path in 0.6.4, and I don't think the merkle proof verification algorithm will work on arrays which have empty data at the end of the inputs, so I'm not sure how it would be possible to eliminate resizing the array.